### ISSUE TITLE
Rust and C# Interop

### ISSUE THREAD

**Issue by dtulepov** [microsoft/CsWinRT#493]
Hi!
Can this project used for cross language interoperability ?
I need consume some rust crate in C#. At this time I'm compiled rust crate that uses that thrid party crate with #[no_mangle] pub extern "C" functions that return C struct, all works ok. Struct contains *mut c_char allocated in rust that then consumed in .net core with DllImport as IntPtr field in struct and converted to string using [Marshal.PtrToStringUTF8](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.ptrtostringutf8?view=netcore-3.1). I call free in Dispose to free string allocated in rust by back then to rust dll like in this article http://jakegoulding.com/rust-ffi-omnibus/string_return/
Can this project simplify cross language interoperability (rust and C#) ?

**Comment by raffaeler** [comment]
Yes, the idea of these project is x-languages interoperability as for [this example](https://github.com/microsoft/CsWinRT/tree/master/Samples/Net5ProjectionSample).
Basically you should create a component with Rust that adhere to the WinRT ABI. Once you have that you can use CsWinRT to generate the interop layer thanks to the component metadata.

CsWinRT is part of a [wider initiative called xlang](https://github.com/Microsoft/xlang/) that includes [C++](https://github.com/microsoft/cppwinrt) and [Rust](https://github.com/microsoft/winrt-rs) projections. There is also one for [Python](https://github.com/Microsoft/xlang/tree/master/src/package/pywinrt/tool) but I don't know its status. You may try to ask in the xlang repo.

HTH

**Comment by dtulepov** [comment]
Thanks! 
Components must be written only in C++ ? Can't find component samples in [Rust projections project](https://github.com/microsoft/winrt-rs) - searched for [DllGetActivationFactory](https://github.com/microsoft/winrt-rs/search?q=DllGetActivationFactory) - only client calls.
Found issue https://github.com/microsoft/winrt-rs/issues/81:
> @kylone This crate already implements the WinRT ABI to the extent that you can, as a _consumer_ or caller, use WinRT components via this binary interface. This issue is about extending that to also act as the _producer_ or callee of such WinRT components by implementing the other side of this binary interface. 
As I understand for now winrt-rs can't create WinRT components, only consume other WinRT components.

**Comment by raffaeler** [comment]
Let's look at the big picture.
WinRT is first of all an ABI (Application Binary Interface) which defines a binary format, a type system and all the metadata required to understand how data is exchanged as well as the threading implications. Nothing new of course, these were the principles of COM and Corba which (both) used IDL (Interface Definition Language) to describe the metadata needed for the interoperability.
The problem with languages like C++ is missing a standard for the binary contract. If you compile something in C++, the binary must be consumed with the same brand/version of the compiler to ensure the compatibility at binary level.
The problem with languages like C and Rust (which AFAIK only uses the C exports) is that the binary contract is too weak. Just a bunch of predefined types and stack conventions but nothing about the component lifecycle and all the other stuff I mentioned before.
WinRT is just the evolution of COM, not really tied to Windows (even if the current implementation is) but with a modern type system, the same of .NET. To be more precise the metadata is the one defined in ECMA-335.

To sum up, any language that is able to adhere to the WinRT contract, can be consumed from the generators like CsWinRT and CppWinRT. They generate 'projections' which projects the usage of those componets in a way that is natural for the language. For example a WinRT object is not created but activated using a factory (CreateInstance). But the C++ and C# projections map the "new" operator to that factory, so that the component appear as it was created in C++ or C#.

With regards to Rust, the project is not finished yet. But if you are able to write a thin layer in C++ proxying Rust, you can just use that for the moment. I believe this is the easiest way for now.

HTH

**Comment by AdamBraden** [comment]
CsWinRT is working on Authoring and Hosting support for creating C# Windows Runtime Components, that will allow the components to be used in non-.NET environments (C++, Rust, etc). See #309 . Expect to see more post NET5 GA.

**Comment by kennykerr** [comment]
Once https://github.com/microsoft/winrt-rs/issues/81 is complete you will be able to author a WinRT component in Rust and call it from C# (or any other language). In the mean time, [C++](https://github.com/microsoft/cppwinrt) is your best bet for authoring WinRT components today.
